ConsoleApplication2/ConsoleApplication2/ConsoleApplication2.cpp

```diff
@@ -0,0 +1,259 @@
+    }
+
+    size_t capacity() {
+        return this->cap;
+    }
+
+    SString& operator= (const SString& in) {
+        this->len = in.len;
+        this->cap = 0;
+
+        reallocStorage(this->len + 1);
+
+        copyString(this->str, in.str, this->len + 1);
+
+        return *this;
+    }
+
+    SString operator+ (const SString& s2) {
+        SString res("");
+
+        res.append(this->str);
+        res.append(s2.str);
+
+        return res;
+    }
+
+    SString& operator+= (const SString& s2) {
+        this->append(s2.str);
+
+        return *this;
+    }
+
+    friend std::ostream& operator<< (std::ostream& out, const SString& s) {
+        out << s.str;
+        return out;
+    }
+
+private:
+    const size_t len_alloc = ALLOCATE_BYTES;
+
+    char* str;
+    size_t len; // current length string 
+    size_t cap; // allocated memory 
+
+    void reallocStorage(size_t n = ALLOCATE_BYTES);
+    bool isFreeStorage();
+    void copyString(char* dest, const char* source, size_t len);
+};
+
+void SString::reallocStorage(size_t n) {
+    this->cap += (n / len_alloc) * len_alloc;
+    this->cap += (n % len_alloc) ? len_alloc : 0;
+   
+    if (!this->str) {
+        std::cout << "memory allocated" << std::endl;
+        this->str = new char[this->cap];
+    }
+    else {
+        // alloc memory + create new pointer 
+        char* str_new = new char[this->cap];
+        // copy data
+        copyString(str_new, this->str, this->len);
+
+        // delete old pointer
+        delete[] this->str;
+        this->str = str_new;
+    }
+}
+
+bool SString::isFreeStorage() {
+    return (this->cap > this->len + 1);
+}
+
+void SString::copyString(char* dest, const char* source, size_t len) {
+    if (!dest || !source || !len)
+        return;
+
+    for (size_t i = 0; i < len; i++) {
+        dest[i] = source[i];
+    }
+}
+
+void SString::append(const char* substr) {
+    if (!substr) // throw exeption
+        return;
+
+    while (*substr != '\0') {
+        if (!isFreeStorage()) {
+            reallocStorage();
+        }
+        this->str[this->len++] = *substr++;
+    }
+    this->str[this->len] = '\0';
+}
+
+void SString::append(const char* substr, size_t sublen) {
+    if (!substr) // throw exeption
+        return;
+
+    while (sublen-- > 0 && *substr != '\0') {
+        if (!isFreeStorage()) {
+            reallocStorage();
+        }
+        this->str[this->len++] = *substr++;
+    }
+    this->str[this->len] = '\0';
+}
+
+void SString::append(const char* in, index_t start, index_t finish) {
+    for (index_t i = start; i < finish; i++) {
+        if (!isFreeStorage()) {
+            reallocStorage();
+        }
+        this->str[this->len++] = *in++;
+    }
+    this->str[this->len] = '\0';
+}
+
+char& SString::at(index_t ind) {
+    try {
+        if (ind < this->len)
+            return this->str[ind];
+
+        throw std::exception("undefined index");
+    } 
+    catch (const std::exception& err) {

```

Comment:
No need to catch exception here. This should go to the user of this class so he knows that sth went wrong in his code.